routing/http/server/filters.go

```diff
@@ -0,0 +1,201 @@
+			peerRecord = applyFilters(peerRecord, filterAddrs, filterProtocols)
+			if peerRecord == nil {
+				return iter.Result[types.Record]{Err: errors.New("record is nil")}
+			}
+			v.Val = peerRecord
+		}
+		return v
+	})
+
+	// filter out nil results and errors
+	filteredIter := iter.Filter(mappedIter, func(v iter.Result[types.Record]) bool {
+		return v.Err == nil && v.Val != nil
+	})
+
+	return filteredIter
+}
+
+//lint:ignore U1000 // ignore unused
+func filterRecords(records []types.Record, filterAddrs, filterProtocols []string) []types.Record {
+	if len(filterAddrs) == 0 && len(filterProtocols) == 0 {
+		return records
+	}
+
+	filtered := make([]types.Record, 0, len(records))
+
+	for _, record := range records {
+		// TODO: Handle SchemaBitswap
+		if schema := record.GetSchema(); schema == types.SchemaPeer {
+			peer, ok := record.(*types.PeerRecord)
+			if !ok {
+				logger.Errorw("problem casting find providers result", "Schema", record.GetSchema(), "Type", reflect.TypeOf(record).String())
+				// if the type assertion fails, we exlude record from results
+				continue
+			}
+
+			record := applyFilters(peer, filterAddrs, filterProtocols)
+
+			if record != nil {
+				filtered = append(filtered, record)
+			}
+
+		} else {
+			// Will we ever encounter the SchemaBitswap type? Evidence seems to suggest that no longer
+			logger.Errorw("encountered unknown provider schema", "Schema", record.GetSchema(), "Type", reflect.TypeOf(record).String())
+		}
+	}
+	return filtered
+}
+
+// Applies the filters. Returns nil if the provider does not pass the protocols filter
+// The address filter is more complicated because it potentially modifies the Addrs slice.
+func applyFilters(provider *types.PeerRecord, filterAddrs, filterProtocols []string) *types.PeerRecord {
+	if len(filterAddrs) == 0 && len(filterProtocols) == 0 {
+		return provider
+	}
+
+	if !applyProtocolFilter(provider.Protocols, filterProtocols) {
+		// If the provider doesn't match any of the passed protocols, the provider is omitted from the response.
+		return nil
+	}
+
+	// return untouched if there's no filter or filterAddrsQuery contains "unknown" and provider has no addrs
+	if len(filterAddrs) == 0 || (len(provider.Addrs) == 0 && slices.Contains(filterAddrs, "unknown")) {
+		return provider
+	}
+
+	filteredAddrs := applyAddrFilter(provider.Addrs, filterAddrs)
+
+	// If filtering resulted in no addrs, omit the provider
+	if len(filteredAddrs) == 0 {
+		return nil
+	}
+
+	provider.Addrs = filteredAddrs
+	return provider
+}
+
+// If there are only negative filters, no addresses will be included in the result. The function will return an empty list.
+// For an address to be included, it must pass all negative filters AND match at least one positive filter.
+func applyAddrFilter(addrs []types.Multiaddr, filterAddrsQuery []string) []types.Multiaddr {
+	if len(filterAddrsQuery) == 0 {
+		return addrs
+	}
+
+	filteredAddrs := make([]types.Multiaddr, 0, len(addrs))
+
```

Comment:
In the case where there are filters, it is likely that many will be filtered out, so probably better not to pre-allocate.
```suggestion
	var filteredAddrs []types.Multiaddr

```